doc-bridgepoint/notes/9706_polys/9706_polys_dnt.md

```diff
@@ -0,0 +1,244 @@
+valid MASL  
+4.2.3 An upgrade strategy shall be considered for models that follow the current
+MASL idiom  
+4.3 An issue shall be raised to analyze the impact on Verifier  
+4.4 An issue shall be raised to analyze the impact on the model compilers  
+
+### 5. Analysis
+
+5.1 Overview
+
+Study the following portion of the "State Machine subsystem:
+![events.png](events.png)
+
+Currently BridgePoint supports abstract polymorphic events by creating an event
+instance in the supertype state machine which has an instance of `SM_PEVT` as
+its subtype. In the subtype state machine, an instance of `SM_EVT`, `SM_SEVT`
+and `SM_NLEVT` are created when the polymorphic event is assigned to a
+transition. The `SM_NLEVT` instance is related to the `SM_PEVT` instance in the
+supertype state machine; the `SM_SEVT` instance participates in the state event
+matrix via `SM_SEME`. Notice that `SM_LEVT` relates to `SM_CRTXN` across R509.
+This means that non-local events (i.e. references to polymorphic events) cannot
+be used in creation transitions. Note also that the relationship between
+`SM_NLEVT` and `SM_PEVT` (R527) does not prevent non-local events from being in
+the same state machine as the polymorphic event even though this behavior is
+enforced in OAL.
+
+The current model has everything necessary to represent concrete polymorphic
+events. The UI could be loosened to allow assigning polymorphic events in
+multiple levels of the subtype/supertype hierarchy. At the supertype level, two
+events would have to exist. One would be the polymorphic event itself, and the
+other would be the non-local state event matrix event that points to the
+polymorphic event. Some masking in the UI would have to be done to prevent both
+events in the supertype from being displayed.
+
+5.2 Vocab
+
+For the rest of this note, the term "local polymorphic event" will be used to
+refer to a polymorphic event that an instance of `SM_NLEVT` refers to which is
+in the same state machine as the non-local event instance (i.e. a concrete
+polymorphic event used in the supertype).
+
+5.3 All events are polymorphic events
+
+In the "concrete polymorphic" idiom, all events are polymorphic -- that is to
+say, an event _must_ be handled in every instance state machine in a
+subtype/supertype hierarchy. Because of this, a column in the state event matrix
+must be created immediately if a supertype event is created. The column will
+contain initial values for the cells ("Can't Happen" or "Ignore"), but these can
+be changed or assigned to a transition. The current BridgePoint behavior is to
+only create the column of the state event matrix when a polymorphic event is
+actually assigned to a transition. This new behavior must be invoked to sync the
+columns at the following actions:
+
+5.3.1 Create event  
+5.3.2 Paste event  
+5.3.3 Create state machine in subtype  
+5.3.4 Add subtype to subtype/supertype hierarchy  
+5.3.5 Unassign event from creation transition (see section 5.5)  
+
```

Comment:
Another use case: delete association somewhere in the sub/super tree.  In this case we would need to remove the SEM entries for polys that are no longer valid.

Reply:
This use case is already handled by the tool and doesn't need special syncing